File: 931. Minimum Falling Path Sum/931. Minimum Falling Path Sum top-down.py
from functools import lru_cache
from typing import List


class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        m = len(matrix)
        n = len(matrix[0])

        @lru_cache(None)
        def dp(row: int, column: int) -> int:
            ways = matrix[row][column]
            if row == 0:
                return ways
            if column > 0 and column < n-1:
                ways += min(
                    dp(row-1, column-1), dp(row-1, column), dp(row-1, column+1))
            elif column > 0:
                ways += min(
                    dp(row-1, column-1), dp(row-1, column))
            elif column < n-1:
                ways += min(
                    dp(row-1, column), dp(row-1, column+1))
            return ways

        return min(dp(m-1, i) for i in range(n))

File: 931. Minimum Falling Path Sum/931. Minimum Falling Path Sum top-down.java
class Solution {
    private int[][] memo, matrix;
    private int m, n;

    private int dp(int row, int column) {
        int ways = matrix[row][column];
        if (row == 0) {
            return ways;
        }
        if (memo[row][column] == 0) {
            if (column > 0 && column < n - 1) {
                ways += Math.min(
                        dp(row - 1, column - 1), Math.min(dp(row - 1, column), dp(row - 1, column + 1)));
            } else if (column > 0) {
                ways += Math.min(
                        dp(row - 1, column - 1), dp(row - 1, column));
            } else if (column < n - 1) {
                ways += Math.min(
                        dp(row - 1, column), dp(row - 1, column + 1));
            }
            memo[row][column] = ways;
        }
        return memo[row][column];
    }

    public int minFallingPathSum(int[][] matrix) {
        this.matrix = matrix;
        this.m = matrix.length;
        this.n = matrix[0].length;
        this.memo = new int[m][n];
        for (int i = 0; i < n; i++) {
            memo[0][i] = matrix[0][i];
        }
        int min = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            min = Math.min(min, dp(m - 1, i));
        }
        return min;
    }
}